x-pack/platform/plugins/shared/observability_ai_assistant/server/service/anonymization/index.ts

```diff
@@ -0,0 +1,288 @@
+const DEFAULT_MAX_CONCURRENT_REQUESTS = 5;
+export interface Dependencies {
+  esClient: {
+    asCurrentUser: ElasticsearchClient;
+  };
+  logger: Logger;
+  anonymizationRules: AnonymizationRule[];
+}
+
+export class AnonymizationService {
+  private readonly esClient: { asCurrentUser: ElasticsearchClient };
+  private readonly logger: Logger;
+  private rules: AnonymizationRule[];
+
+  private currentHashMap: Map<
+    string,
+    { value: string; class_name: string; type: DetectedEntityType }
+  > = new Map();
+
+  constructor({ esClient, logger, anonymizationRules }: Dependencies) {
+    this.esClient = esClient;
+    this.logger = logger;
+    this.rules = anonymizationRules;
+  }
+
+  private async detectNamedEntities(chunks: InferenceChunk[]): Promise<DetectedEntity[]> {
+    this.logger.debug(`Detecting named entities in ${chunks.length} text chunks`);
+
+    // Maximum number of concurrent requests to the ML model
+    const limiter = pLimit(DEFAULT_MAX_CONCURRENT_REQUESTS);
+
+    // Batch size - number of documents to send in each request
+    const BATCH_SIZE = 10;
+
+    // Create batches of chunks for the inference request
+    const batches = chunk(chunks, BATCH_SIZE);
+    this.logger.debug(`Processing ${batches.length} batches of up to ${BATCH_SIZE} chunks each`);
+
+    const tasks = batches.map((batchChunks) =>
+      limiter(async () =>
+        withInferenceSpan('infer_ner', async () => {
+          let response;
+          try {
+            response = await this.esClient.asCurrentUser.ml.inferTrainedModel({
+              model_id: NER_MODEL_ID,
+              docs: batchChunks.map((batchChunk) => ({ text_field: batchChunk.chunkText })),
+            });
+          } catch (error) {
+            throw new Error('NER inference failed', { cause: error });
+          }
+          // Process results from all documents in the batch
+          const batchResults: DetectedEntity[] = [];
+          const inferenceResults = response?.inference_results || [];
+
+          if (inferenceResults.length !== batchChunks.length) {
+            this.logger.warn(
+              `NER returned ${inferenceResults.length} results for ${batchChunks.length} docs in batch`
+            );
+          }
+
+          // Match results with their original chunks to maintain offsets
+          inferenceResults.forEach((result, index) => {
+            const batchChunk = batchChunks[index];
+            const entities = result.entities || [];
+
+            batchResults.push(
+              ...entities.map((e) => ({
+                ...e,
+                start_pos: e.start_pos + batchChunk.charStartOffset,
+                end_pos: e.end_pos + batchChunk.charStartOffset,
+                type: 'ner' as const,
+                hash: getEntityHash(e.entity, e.class_name),
+              }))
+            );
+          });
+
+          return batchResults;
+        })
+      )
+    );
+
+    const results = await Promise.all(tasks);
```

Comment:
What if one of the the tasks failed but the rest succeeds? Do we want to reject it completely like here? If not, maybe we can use something like allSettled here.